File: day24/day24.kt
import java.io.File

class Floor(input: List<String>) {
    data class Tile(val x: Int, val y: Int)
    val tiles = mutableSetOf<Tile>()
    val dir = mapOf(
        "e" to Pair(1, 0),
        "w" to Pair(-1, 0),
        "ne" to Pair(0, 1),
        "nw" to Pair(-1, 1),
        "se" to Pair(1, -1),
        "sw" to Pair(0, -1),
    )
    
    init {
        input.map { parse(it) }.forEach {
            if (it !in tiles) tiles.add(it)
            else tiles.remove(it)
        }
    }

    fun parse(line: String): Tile {
        var p = 0
        var x = 0; var y = 0
        while (p < line.length) {
            val single = dir.contains(line[p].toString())
            val key = line.substring(p, p + if (single) 1 else 2)
            p += key.length
            x += dir[key]!!.first
            y += dir[key]!!.second
        }
        return Tile(x, y)
    }
    
    fun move() {
        val state = tiles.toSet()
        val white = mutableSetOf<Tile>()
        for (b in state) {
            val a = dir.values.map { (dx, dy) -> Tile(b.x + dx, b.y + dy) }
            val n = a.count { it in state }
            if (n == 0 || n > 2) tiles.remove(b)
            white.addAll(a)
        }
        for (w in white.sortedWith { a, b -> if (a.x != b.x) a.x - b.x else a.y - b.y }) {
            val a = dir.values.map { (dx, dy) -> Tile(w.x + dx, w.y + dy) }
            val n = a.count { it in state }
            if (n == 2) tiles.add(w)
        }
    }
}

fun main() {
    val input = File("day24.txt").readLines()
    val floor = Floor(input)
    println(floor.tiles.size)
    (1..100).forEach { floor.move() }
    println(floor.tiles.size)
}

File: 2020/day24/day24.py
moves = {
    'e': (1, 0),
    'w': (-1, 0),
    'ne': (0, 1),
    'nw': (-1, 1),
    'se': (1, -1),
    'sw': (0, -1),
}

def parse(line):
    res = [(0, 0)]
    while line:
        key = line[0] if line[0] in moves else line[:2]
        dx, dy = moves[key]
        res.append((res[-1][0] + dx, res[-1][1] + dy))
        line = line[len(key):]
    return res[-1]

def move(tiles):
    res, adj = tiles.copy(), set()
    for b in tiles:
        a = [(b[0] + dx, b[1] + dy) for dx, dy in moves.values()]
        n = sum(x in tiles for x in a)
        if n == 0 or n > 2: res.remove(b)
        adj.update(a)
    for w in adj:
        a = [(w[0] + dx, w[1] + dy) for dx, dy in moves.values()]
        n = sum(x in tiles for x in a)
        if n == 2: res.add(w)
    return res

input_ = [x.rstrip() for x in open('day24.txt')]
tiles = set()
for line in input_:
    pos = parse(line)
    if pos not in tiles:
        tiles.add(pos)
    else:
        tiles.remove(pos)
print(len(tiles))

for _ in range(100):
    tiles = move(tiles)
print(len(tiles))
